### ISSUE TITLE
Debugging with the RequestInspector and ordering

### ISSUE THREAD

**Issue by lawrencegripper** [Azure/go-autorest#252]
Hi, 

I'm currently working on an issue I've been seeing talking to an Azure service through `azure-sdk-for-go` and `go-autorest`

The `RequestInspector` property looked to be my savior but it isn't currently called as the last step so the authentication token isn't written into the request yet. I expected this to be the final step before sending a request and wanted to use this to see every aspect of the request before it was sent. 

I'd propose altering the ordering of the `prepare` statements [here](https://github.com/Azure/go-autorest/blob/master/autorest/client.go#L207) to something like:

```golang
	r, err := Prepare(r,
		c.WithAuthorization())
	if err != nil {
		var resp *http.Response
		if detErr, ok := err.(DetailedError); ok {
			// if the authorization failed (e.g. invalid credentials) there will
			// be a response associated with the error, be sure to return it.
			resp = detErr.Response
		}
		return resp, NewErrorWithError(err, "autorest/Client", "Do", nil, "Preparing request failed")
	}

	r, _ = Prepare(r,
		c.WithInspection())
```

This ensured my inspector (below) dumped the request as it was going to be sent 

```golang
func LogRequestPreparer() autorest.PrepareDecorator {
	return func(p autorest.Preparer) autorest.Preparer {
		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
			resDump, _ := httputil.DumpRequestOut(r, true)
			log.Println(string(resDump))
			return r, nil
		})
	}
}
```

What do you think?

**Comment by jhendrixMSFT** [comment]
Thanks for reporting this.  There are two issues here.  First is that in your preparer you must call prepare on the passed in preparer before executing your logging code.
```golang
func LogRequestPreparer() autorest.PrepareDecorator {
	return func(p autorest.Preparer) autorest.Preparer {
		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
			r, err := p.Prepare(r)
			if err == nil {
				resDump, _ := httputil.DumpRequestOut(r, true)
				log.Println(string(resDump))
			}
			return r, nil
		})
	}
}
```
The reason being that the preparers form a linked list of operations, and calling it first like this allows them to execute in declared order.
The second issue is that we have a bug in go-autorest, in Client.Do() we should be calling c.WithInspection() last so that it operates on the final request (I will fix this today).
With these two fixes your logger works as expected and dumps the authorization header to the console.

**Comment by jhendrixMSFT** [comment]
Fix is in v10.1.3
https://github.com/Azure/go-autorest/releases/tag/v10.1.3
